api/src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java

```diff
@@ -0,0 +1,200 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.flow;
+
+import hudson.AbortException;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.FilePath;
+import hudson.Util;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.org.apache.tools.tar.TarInputStream;
+import hudson.util.DirScanner;
+import hudson.util.io.ArchiverFactory;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map;
+import java.util.TreeMap;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import jenkins.model.ArtifactManager;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.tools.tar.TarEntry;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+
+/**
+ * Manages per-build stashes of files.
+ * Unlike artifacts managed by {@link ArtifactManager}, stashes:
+ * <ul>
+ * <li>Are expected to be transferred to other workspaces during the build.
+ * <li>Generally are discarded when the build finishes.
+ * <li>Are not exposed as part of the build outside Jenkins, for example via REST.
+ * <li>Are stored in an archive format with a simple name, not necessarily related to filenames.
+ * </ul>
+ */
+public class StashManager {
+
+    /**
+     * Saves a stash of some files from a build.
+     * @param build a build to use as storage
+     * @param name a simple name to assign to the stash (must follow {@link Jenkins#checkGoodName} constraints)
+     * @param workspace a directory to use as a base
+     * @param includes a set of Ant-style file includes, separated by commas; null/blank is allowed as a synonym for {@code **} (i.e., everything)
+     * @param excludes an optional set of Ant-style file excludes
+     */
+    public static void stash(@Nonnull Run<?,?> build, @Nonnull String name, @Nonnull FilePath workspace, @Nonnull TaskListener listener, @CheckForNull String includes, @CheckForNull String excludes) throws IOException, InterruptedException {
+        Jenkins.checkGoodName(name);
+        File storage = storage(build, name);
```

Comment:
should this complain if it already exists? it looks like now it silently overwrites (which may be correct - but it doesn't mention this on the listener so could cause confusion if a user cuts and pastes or uses some library function multiple times without thinking?